### PR TITLE
feature: Support two abort mode. `cancel` to abort the pending task, …

### PR DESCRIPTION
**Pull request by JetDrag**
feature:  Support two abort mode. `cancel` to abort the pending task, and the default `abort` to abort the pending or running task.

### CODE REVIEW

**Comment by JetDrag** [comment]
Hi, @isra17 I just open a new clean PR for review.

File: src/dramatiq_abort/middleware.py [hunk 1]
```diff
@@ -137,15 +161,21 @@ def _watcher(self) -> None:
...
+    def id_to_key(message_id: str, mode: AbortMode = AbortMode.ABORT) -> bytes:
+        return (
+            ("abort:" if mode is AbortMode.ABORT else "cancel:") + message_id
+        ).encode()
 
     @staticmethod
-    def key_to_id(key: bytes) -> str:
-        return key.decode()[6:]
+    def key_to_id(key: bytes, mode: AbortMode = AbortMode.ABORT) -> str:
+        return key.decode()[6:] if mode is AbortMode.ABORT else key.decode()[7:]
```
Comment (isra17) [inline src/dramatiq_abort/middleware.py:171]: `return key.split(b":")[-1].decode()`

**Review by isra17 — CHANGES_REQUESTED** [review]
Looking good overall, thanks for your work!

Few more comments, and maybe rebase on master too, there will be conflict with another change to `abort` parameters.

On the merge, that would be great if you keep all extra parameters as `kwargs` only:

`abort(message_id, *, ttl, mode)`

Thanks again!

**Comment by JetDrag** [comment]
Hi @isra17, do you have time to review the change and release a new version? So we don't have to deploy a custom version in out production environment.

File: src/dramatiq_abort/backend.py [hunk 2]
```diff
@@ -36,8 +36,8 @@ def poll(self, key: bytes) -> bool:  # pragma: no cover
         raise NotImplementedError
 
     @abc.abstractmethod
-    def notify(self, key: bytes, ttl: int) -> None:  # pragma: no cover
-        """Signal an event.
+    def notify(self, keys: Iterable[bytes], ttl: int) -> None:  # pragma: no cover
```
Comment (isra17) [inline src/dramatiq_abort/backend.py:39]: Why are you changing this? Looking at the implementation, `notify` appears to be the same as `notify_many`.

File: src/dramatiq_abort/middleware.py [hunk 3]
```diff
@@ -126,22 +139,32 @@ def after_process_message(
 
     after_skip_message = after_process_message
 
-    def abort(self, message_id: str, abort_ttl: Optional[int] = None) -> None:
+    def abort(
+        self,
+        message_id: str,
+        mode: AbortMode = AbortMode.ABORT,
```
Comment (isra17) [inline src/dramatiq_abort/middleware.py:145]: Please add `mode` after `abort_ttl` to keep this backward compatible.

File: src/dramatiq_abort/middleware.py [hunk 4]
```diff
@@ -126,22 +139,32 @@ def after_process_message(
...
             time.sleep(self.wait_timeout / 1000)
             return
 
-        abort_keys = [self.id_to_key(id_) for id_ in message_ids]
+        abort_keys = [self.id_to_key(id_, AbortMode.ABORT) for id_ in message_ids]
         key = self.backend.wait_many(abort_keys, self.wait_timeout)
         if not key:
             return
+        else:
+            self.backend.poll(key)
```
Comment (isra17) [inline src/dramatiq_abort/middleware.py:167]: Why is this needed? `poll` has the same side-effects than `wait_many`

File: src/dramatiq_abort/middleware.py [hunk 5]
```diff
@@ -157,17 +180,19 @@ def _watcher(self) -> None:
...
     @staticmethod
-    def id_to_key(message_id: str) -> bytes:
-        return ("abort:" + message_id).encode()
+    def id_to_key(message_id: str, mode: AbortMode = AbortMode.ABORT) -> bytes:
+        return (mode.value + ":" + message_id).encode()
 
     @staticmethod
     def key_to_id(key: bytes) -> str:
-        return key.decode()[6:]
+        return key.split(b":")[-1].decode()
```
Comment (isra17) [inline src/dramatiq_abort/middleware.py:188]: Just for safety, let's use `split(b":", 1)`. This way an id with `:` will still work.

File: tests/test_abortable.py [hunk 6]
```diff
@@ -103,8 +162,8 @@ def do_work() -> None:
 
     # If I send it a message
     message = do_work.send()
-    # And abort right after.
-    abort(message.message_id)
+    # And cancel right after.
+    abort(message.message_id, mode=AbortMode.CANCEL)
```
Comment (isra17) [inline tests/test_abortable.py:166]: `mode=mode`?

File: src/dramatiq_abort/backend.py (hunk 2 above)
Comment (isra17) [reply src/dramatiq_abort/backend.py:39]: I would suggest to get rid of `notify_many` and only keep `notify` with your changes. Also, please update the docstring for `key`!

**Review by isra17 — CHANGES_REQUESTED** [review]
Few small changes or nitpick, I think that should be the last round of review!

Thanks again

File: src/dramatiq_abort/middleware.py (hunk 4 above)
Comment (JetDrag) [reply src/dramatiq_abort/middleware.py:167]: Sorry, It's the omitted unused code during the recent review.

**Comment by JetDrag** [comment]
Hi, @isra17, I've resolved the all discussion. Thanks for your patient, review again please.

File: src/dramatiq_abort/middleware.py (hunk 3 above)
Comment (isra17) [reply src/dramatiq_abort/middleware.py:145]: This wasn't addressed!

it should be:

```python
    def abort(
        self,
        message_id: str,
        abort_ttl: Optional[int] = None,
        mode: AbortMode = AbortMode.ABORT,
    ) -> None
```

Because the keyward arguments are not non-position (`*, abort_ttl...)`, some code might be doing `abort(message_id, 600)`. With your current change, this working code will break.

Comment (isra17) [reply src/dramatiq_abort/middleware.py:145]: We can also do like you did below and add the non-positional marker (`*`). At least this will allow linter to catch the issue.

**Review by isra17 — CHANGES_REQUESTED** [review]
One very last comment, then I merge and release!

File: src/dramatiq_abort/middleware.py (hunk 3 above)
Comment (JetDrag) [reply src/dramatiq_abort/middleware.py]: I just remove the '*' to keep the compatibility and take the correct args order.

**Comment by JetDrag** [comment]
I've resolved the recent discussion，review again pls.

**Comment by isra17** [comment]
Thank you for your work!

**Comment by isra17** [comment]
This is released in `0.4.0`!

**Comment by JetDrag** [comment]
Thank you @isra17

**Comment by JetDrag** [comment]
But there is still a problem that the doc is not up to date anywhere. It seems that the doc is at version 0.1.x. @isra17

**Comment by isra17** [comment]
Fixed the github action, should work now
